common/beerocks/bcl/include/bcl/beerocks_socket_event_loop.h

```diff
@@ -0,0 +1,128 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _BEEROCKS_SOCKET_EVENT_LOOP_H_
+#define _BEEROCKS_SOCKET_EVENT_LOOP_H_
+
+#include "beerocks_event_loop.h"
+#include "network/socket.h"
+
+#include <memory>
+#include <unordered_map>
+
+namespace beerocks {
+
+/**
+ * @brief ePoll based implementation of the EventLoop interface.
+ * @see EventLoop
+ * 
+ * This class uses the Linux epoll APIs for monitoring the provided sockets for I/O operations.
+ * Timeout operations are achieved by using the timerfd mechanism, which delivers timer 
+ * expiration notifications via a file descriptor.
+ */
+class SocketEventLoop : public EventLoop<std::shared_ptr<Socket>> {
+public:
+    /**
+     * @brief Class constructor.
+     * 
+     * Initializes an epoll file descriptor.
+     * 
+     * @param [in] timeout Sets the master timeout (in milliseconds) for the event loop.
+     */
+    explicit SocketEventLoop(std::chrono::milliseconds timeout = std::chrono::milliseconds::min());
+
+    /**
+     * @brief Class destructor.
+     */
+    virtual ~SocketEventLoop();
+
+    /**
+     * @see EventPoll::add_socket
+     */
+    virtual bool
+    add_event(SocketEventLoop::EventType socket, const SocketEventLoop::EventHandlers &handlers,
```

Comment:
This works as well (I think) and is clearer IMO:
```suggestion
    add_event(std::shared_ptr<Socker> socket, const EventHandlers &handlers,
```

However, that is really weird. The shared_ptr indicates that the caller may still hold a reference to the socket after this call, but may also discard its reference. However, the handlers are passed as const ref, so there the caller *must* still hold its reference until the event is removed again. I think both should be made consistent, i.e. either both of them are shared_ptr, or both of them are reference. I tend to prefer both const-ref, i.e. it's the caller's responsibility to make sure the socket stays alive until it is removed from the even loop with del_event(). Since del_event takes the socket as an argument, that's kind of obvious :-)

Actually, with this in mind, and given how EventHandlers is structured, wouldn't it make more sense to make the socket a member of EventHandlers? Then you will guarantee consistent ownership for both.

Reply:
Passing the event handlers a const reference is a mistake. The handlers are copied and stored in the event loop's internal data structure (i.e. `EventData`).
The socket on the other hand should be kept alive by the caller until it's removed from the loop, so passing a `shared_ptr` to it does make sense.

As for the `EventType` and `EventLoopType` you think it's confusing?
My thought was that defining the type once (i.e. `class SocketEventLoop : public EventLoop<std::shared_ptr<Socket>>`) and then using the "Type" definitions should simplify things, but perhaps it just makes things less clear...